### PR TITLE
Update the name of a product variation to include all attributes with "Any" value

### LINKED ISSUE #2595: Update the name of a product variation to include value of all attributes (including "Any")
Ref p91TBi-33Q-p2#comment-1383

Depends on https://github.com/woocommerce/woocommerce-ios/issues/2575 so that we can also pass a variation's parent product's attributes for the name of a variation

Right now, we just show the non-empty attributes (where the value is not "Any") for a variation's name. When a variation has no non-empty attributes, then the name is empty. Let's update a variation's name to include "Any" for all possible attributes.

screen-shot-2020-08-05-at-9 44 07-pm

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #2595 

## Changes

- In `EditableProductVariationModel`, added a dependency `allAttributes: [ProductAttribute]` for generating the variation name lazily considering all attributes from its parent product
- In `ProductVariationsViewController`, updated the initializer to pass in the variable product `Product` instead of just the site ID and product ID so that we can also initialize the `allAttributes: [ProductAttribute]` from the variable product. Updated its cell configuration to use `EditableProductVariationModel` instead of `ProductVariation`, and updated `ProductsTabProductViewModel`'s variation initializer to take in `EditableProductVariationModel` for its generated name.
- Also DI'ed `allAttributes: [ProductAttribute]` to `ProductVariationFormViewModel` for updating `EditableProductVariationModel` on field changes
- Added a new case `ProductFormEditAction.variationName` and updated `ProductVariationFormActionsFactory` to use `.variationName` instead of `.name` for the name row
- Added the corresponding new case `ProductFormSection.PrimaryFieldRow.variationName(name: String)` for the variation name row. Configured this row to use `BasicTableViewCell` with readonly name that has the same style as product name row.
- Removed `isEditable` associated value from `ProductFormSection.PrimaryFieldRow.name` now that the readonly variation name is in a separate case
- Added unit tests for the variation name

## Testing

Prerequisite: the site has a variation with 2 types of variations:
1. Variations with all the attributes specified (e.g. there is an option for all the attributes)
2. Variations with some or all attributes as "Any" in core

- Go to the Products tab
- Tap on a variable product
- Tap on the variations row --> the variation list should show each variation name as in design in #2595 screenshot ("Any {attribute}" or the option value for each attribute, sorted by attribute position)
- Tap on a variation --> the variation name should be the same as in variation list

## Example screenshots

mode | variation list | variation
-- | -- | --
Light | Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 19 | Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 28
Dark | Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 48 | Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 50

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift [hunk 1]
```diff
@@ -4,8 +4,28 @@ import Yosemite
...
+            .sorted(by: { (lhs, rhs) -> Bool in
+                lhs.position < rhs.position
+            })
+            .map { attribute in
+            guard let variationAttribute = variationAttributes.first(where: { $0.id == attribute.attributeID && $0.name == attribute.name }) else {
+                // The variation doesn't have an option set for this attribute, and we show "Any \(attributeName)" in this case.
+                return String.localizedStringWithFormat(Localization.anyAttirubteFormat, attribute.name)
+            }
+            return variationAttribute.option
+        }.joined(separator: " - ")
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift:28]: 👏

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift [hunk 2]
```diff
@@ -133,3 +153,10 @@ extension EditableProductVariationModel: Equatable {
         return lhs.productVariation == rhs.productVariation
     }
 }
+
+extension EditableProductVariationModel {
+    enum Localization {
+        static let anyAttirubteFormat =
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift:159]: Typo:
```suggestion
        static let anyAttributeFormat =
```

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift [hunk 3]
```diff
@@ -4,8 +4,28 @@ import Yosemite
...
+private extension EditableProductVariationModel {
+    func generateName(variationAttributes: [ProductVariationAttribute], allAttributes: [ProductAttribute]) -> String {
+        return allAttributes
+            .sorted(by: { (lhs, rhs) -> Bool in
+                lhs.position < rhs.position
+            })
+            .map { attribute in
+            guard let variationAttribute = variationAttributes.first(where: { $0.id == attribute.attributeID && $0.name == attribute.name }) else {
+                // The variation doesn't have an option set for this attribute, and we show "Any \(attributeName)" in this case.
+                return String.localizedStringWithFormat(Localization.anyAttirubteFormat, attribute.name)
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift:25]:
```suggestion
                return String.localizedStringWithFormat(Localization.anyAttributeFormat, attribute.name)
```

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/EditableProductVariationModelTests.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        ]
+        let variation = MockProductVariation().productVariation().copy(attributes: variationAttributes)
+
+        // Action
+        let name = EditableProductVariationModel(productVariation: variation, allAttributes: allAttributes).name
+
+        // Assert
+        let expectedName = [
+            "Orange",
+            String.localizedStringWithFormat(EditableProductVariationModel.Localization.anyAttirubteFormat, "Brand")
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/EditableProductVariationModelTests.swift:24]:
```suggestion
            String.localizedStringWithFormat(EditableProductVariationModel.Localization.anyAttributeFormat, "Brand")
```

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift (hunk 2 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift]: Oops, my eyes must have been sore at the time 😓 updated here: https://github.com/woocommerce/woocommerce-ios/pull/2608/commits/796493551517b221555ddf8129a956426a194b26

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift (hunk 3 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/EditableProductVariationModel.swift]: updated! https://github.com/woocommerce/woocommerce-ios/pull/2608/commits/796493551517b221555ddf8129a956426a194b26

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/EditableProductVariationModelTests.swift (hunk 4 above)
Comment (jaclync) [reply WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/EditableProductVariationModelTests.swift]: Thanks for catching this Paolo! updated here: https://github.com/woocommerce/woocommerce-ios/pull/2608/commits/796493551517b221555ddf8129a956426a194b26

**Pull request merged**